File: audit_investors.py
import os
import json
import urllib.request
from dotenv import load_dotenv

load_dotenv('frontend/.env')
SUPABASE_URL = 'https://rjdewjyhtbfkujhvkwig.supabase.co'
SUPABASE_KEY = os.environ.get('VITE_SUPABASE_ANON_KEY')

def fetch_all_investors():
    all_investors = []
    limit = 1000
    offset = 0
    
    while True:
        url = f"{SUPABASE_URL}/rest/v1/investors?select=id,name,email,linkedin_url,twitter_url&limit={limit}&offset={offset}"
        req = urllib.request.Request(url, headers={
            'apikey': SUPABASE_KEY,
            'Authorization': f'Bearer {SUPABASE_KEY}',
            'Range-Unit': 'items'
        })
        
        try:
            with urllib.request.urlopen(req) as response:
                data = json.loads(response.read().decode('utf-8'))
                all_investors.extend(data)
                
                if len(data) < limit:
                    break
                offset += limit
        except Exception as e:
            print(f"Error fetching data at offset {offset}: {e}")
            break
            
    return all_investors

def audit():
    print("Fetching investors from database...")
    investors = fetch_all_investors()
    total = len(investors)
    print(f"Total investors fetched: {total}\n")
    
    if total == 0:
        return

    stats = {
        "ideal": 0,
        "missing_email_only": 0,
        "missing_linkedin_only": 0,
        "missing_twitter_only": 0,
        "missing_two_fields": 0,
        "missing_all_three": 0,
        
        # Absolute missing counts
        "total_missing_email": 0,
        "total_missing_linkedin": 0,
        "total_missing_twitter": 0,
    }

    for inv in investors:
        has_email = bool(inv.get('email') and str(inv.get('email')).strip())
        has_li = bool(inv.get('linkedin_url') and str(inv.get('linkedin_url')).strip())
        has_tw = bool(inv.get('twitter_url') and str(inv.get('twitter_url')).strip())
        
        fields_count = sum([has_email, has_li, has_tw])
        
        if not has_email: stats["total_missing_email"] += 1
        if not has_li: stats["total_missing_linkedin"] += 1
        if not has_tw: stats["total_missing_twitter"] += 1

        if fields_count == 3:
            stats["ideal"] += 1
        elif fields_count == 2:
            if not has_email:
                stats["missing_email_only"] += 1
            elif not has_li:
                stats["missing_linkedin_only"] += 1
            elif not has_tw:
                stats["missing_twitter_only"] += 1
        elif fields_count == 1:
            stats["missing_two_fields"] += 1
        else:
            stats["missing_all_three"] += 1

    print("--- АУДИТ БАЗЫ ДАННЫХ ИНВЕСТОРОВ ---")
    print(f"Всего записей: {total}")
    print(f"✅ Идеальные (есть Email, LinkedIn, Twitter): {stats['ideal']} ({(stats['ideal']/total)*100:.1f}%)")
    print("-" * 40)
    print("⚠️ Не хватает ровно одного поля:")
    print(f"  - Нет Email (но есть соцсети): {stats['missing_email_only']}")
    print(f"  - Нет LinkedIn (но есть Email и Twitter): {stats['missing_linkedin_only']}")
    print(f"  - Нет Twitter (но есть Email и LinkedIn): {stats['missing_twitter_only']}")
    print("-" * 40)
    print(f"🚨 Не хватает двух полей: {stats['missing_two_fields']} ({(stats['missing_two_fields']/total)*100:.1f}%)")
    print(f"🗑️ Мусорные (нет ни Email, ни LinkedIn, ни Twitter): {stats['missing_all_three']} ({(stats['missing_all_three']/total)*100:.1f}%)")
    print("-" * 40)
    print("Общая статистика отсутствующих полей:")
    print(f"Всего без Email:    {stats['total_missing_email']} ({(stats['total_missing_email']/total)*100:.1f}%)")
    print(f"Всего без LinkedIn: {stats['total_missing_linkedin']} ({(stats['total_missing_linkedin']/total)*100:.1f}%)")
    print(f"Всего без Twitter:  {stats['total_missing_twitter']} ({(stats['total_missing_twitter']/total)*100:.1f}%)")

if __name__ == "__main__":
    audit()

File: frontend_old/audit_investors.mjs
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';

dotenv.config({ path: 'frontend/.env' });

const supabaseUrl = process.env.VITE_SUPABASE_URL || 'https://rjdewjyhtbfkujhvkwig.supabase.co';
const supabaseAnonKey = process.env.VITE_SUPABASE_ANON_KEY;

const supabase = createClient(supabaseUrl, supabaseAnonKey);

async function audit() {
  console.log("Fetching investors...");
  let all_investors = [];
  let limit = 1000;
  let offset = 0;

  while (true) {
    const { data, error } = await supabase
      .from('investors')
      .select('id, name, email, linkedin_url, twitter_url')
      .range(offset, offset + limit - 1);

    if (error) {
      console.error("Error fetching data:", error);
      break;
    }

    all_investors = all_investors.concat(data);
    if (data.length < limit) break;
    offset += limit;
  }

  const total = all_investors.length;
  console.log(`Total investors fetched: ${total}\n`);

  if (total === 0) return;

  const stats = {
    ideal: 0,
    missing_email_only: 0,
    missing_linkedin_only: 0,
    missing_twitter_only: 0,
    missing_two_fields: 0,
    missing_all_three: 0,
    total_missing_email: 0,
    total_missing_linkedin: 0,
    total_missing_twitter: 0,
  };

  for (const inv of all_investors) {
    const has_email = !!(inv.email && inv.email.trim());
    const has_li = !!(inv.linkedin_url && inv.linkedin_url.trim());
    const has_tw = !!(inv.twitter_url && inv.twitter_url.trim());

    const fields_count = [has_email, has_li, has_tw].filter(Boolean).length;

    if (!has_email) stats.total_missing_email++;
    if (!has_li) stats.total_missing_linkedin++;
    if (!has_tw) stats.total_missing_twitter++;

    if (fields_count === 3) stats.ideal++;
    else if (fields_count === 2) {
      if (!has_email) stats.missing_email_only++;
      else if (!has_li) stats.missing_linkedin_only++;
      else if (!has_tw) stats.missing_twitter_only++;
    } else if (fields_count === 1) {
      stats.missing_two_fields++;
    } else {
      stats.missing_all_three++;
    }
  }

  console.log("--- АУДИТ БАЗЫ ДАННЫХ ИНВЕСТОРОВ ---");
  console.log(`Всего записей: ${total}`);
  console.log(`✅ Идеальные (есть Email, LinkedIn, Twitter): ${stats.ideal} (${(stats.ideal / total * 100).toFixed(1)}%)`);
  console.log("-".repeat(40));
  console.log("⚠️ Не хватает ровно одного поля:");
  console.log(`  - Нет Email (но есть соцсети): ${stats.missing_email_only}`);
  console.log(`  - Нет LinkedIn (но есть Email и Twitter): ${stats.missing_linkedin_only}`);
  console.log(`  - Нет Twitter (но есть Email и LinkedIn): ${stats.missing_twitter_only}`);
  console.log("-".repeat(40));
  console.log(`🚨 Не хватает двух полей: ${stats.missing_two_fields} (${(stats.missing_two_fields / total * 100).toFixed(1)}%)`);
  console.log(`🗑️ Мусорные (нет ни Email, ни LinkedIn, ни Twitter): ${stats.missing_all_three} (${(stats.missing_all_three / total * 100).toFixed(1)}%)`);
  console.log("-".repeat(40));
  console.log("Общая статистика отсутствующих полей:");
  console.log(`Всего без Email:    ${stats.total_missing_email} (${(stats.total_missing_email / total * 100).toFixed(1)}%)`);
  console.log(`Всего без LinkedIn: ${stats.total_missing_linkedin} (${(stats.total_missing_linkedin / total * 100).toFixed(1)}%)`);
  console.log(`Всего без Twitter:  ${stats.total_missing_twitter} (${(stats.total_missing_twitter / total * 100).toFixed(1)}%)`);
}

audit();
